torch/onnx/_internal/exporter.py

```diff
@@ -0,0 +1,198 @@
+    """Options to influence the ONNX exporter."""
+
+    opset_version: Optional[int] = None
+    """opset_version: Target ONNX opset version.
+
+    Defaults to the latest supported ONNX opset version."""
+
+    dynamic_shapes: Optional[bool] = None
+    """Hints shape information for input/output tensors.
+
+    When `True`, all inputs shapes are considered dynamic.
+    When `False`, all inputs shapes are considered static.
+
+    Defaults to the most compatible setting.
+    """
+
+    logger: Optional[logging.Logger] = None
+    """Logging level for the exporter. Defaults to the effective level
+    of the current Python :py:mod:`logging` logger."""
+
+
+class ExportOutputSerializer(Protocol):
+    """Protocol for serializing an ONNX graph into a specific format (e.g. Protobuf)."""
+
+    def serialize(self, export_output: "ExportOutput", destination: io.BytesIO) -> None:
+        pass
+
+
+class ProtobufExportOutputSerializer:
+    """Serializes ONNX graph as Protobuf."""
+
+    def serialize(self, export_output: "ExportOutput", destination: io.BytesIO) -> None:
+        import onnx
+
+        if not isinstance(export_output.model_proto, onnx.ModelProto):
+            raise ValueError("export_output.ModelProto is not an onnx.ModelProto")
+        destination.write(export_output.model_proto.SerializeToString())
+
+
+class ExportOutput:
+    """An in-memory representation of a PyTorch model that has been exported to ONNX."""
+
+    _model_proto: Final[onnx.ModelProto]
+
+    def __init__(self, model_proto: onnx.ModelProto):
+        self._model_proto = model_proto
+
+    @property
+    def model_proto(self) -> onnx.ModelProto:
+        """An onnx.ModelProto. Callers are expected to type and null-check the value."""
+
+        return self._model_proto
+
+    def save(
+        self,
+        destination: Union[str, io.BytesIO],
+        *,
+        serializer: Optional[ExportOutputSerializer] = None,
+    ) -> None:
+        """Saves the in-memory ONNX representation of the model to `destination`
+        using specified `serializer`."""
+
+        if serializer is None:
+            serializer = ProtobufExportOutputSerializer()
+        if isinstance(destination, str):
+            with open(destination, "wb") as f:
+                serializer.serialize(self, f)
+        else:
+            serializer.serialize(self, destination)
+
+
+class Exporter(abc.ABC):
+    def __init__(
+        self,
+        options: ExportOptions,
+        model: torch.nn.Module,
+        model_args: Sequence[Any],
+        model_kwargs: Mapping[str, Any],
+    ):
+        self.options = options
+        self.model = model
+        self.model_args = model_args
+        self.model_kwargs = model_kwargs
+
+    @abc.abstractmethod
+    def run(self) -> ExportOutput:
+        pass
+
+    @property
+    def logger(self) -> logging.Logger:
+        return self.options.logger
+
+
+def _resolve_export_options(options: Optional[ExportOptions]) -> ExportOptions:
```

Comment:
Is the main goal of having `_resolve_export_options` is to not mutate user passed in `ExportOptions` object?